### ISSUE TITLE
Bogus `unnecessary_this` when shadowing inside `switch` statement

### ISSUE THREAD

**Issue by mernen** [dart-lang/linter#4457]
Under certain conditions, creating a shadowing variable inside a `case` body won't be properly detected, causing a `unnecessary_this` warning.

```dart
class C {
  String? name;
  
  void example(bool printName) {
    switch (printName) {
      case true:
        var name = this.name!;
        //         ~~~~~~~~~ Unnecessary 'this.' qualifier.
        print(name);
        break;

      case false:
        break;
    }
  }
}
```

Expected: the qualifier is necessary due to local variable shadowing, and therefore shouldn't produce a lint.

It's apparently not triggered on the last `case` (or `default`). Moving `case false` up in the example above will suppress the lint. Also, it's only triggered if the variable is declared in the case body — splitting the variable declaration and assignment still triggers the lint, but moving the declaration to a higher scope suppresses it.

This appears to be a regression from Dart 2 — testable on dartpad.dev through the "old" channel, currently offering Dart 2.19.6.

Finally, this might be related to #4381, but a cursory look suggested it's a different problem.

**Comment by pq** [comment]
Thanks for reporting this!  It's related to #4381 insofar as I expect the cause is similarly percolating up from `context.resolveNameInScope` in analyzer but the fix could be different.  @scheglov?

**Comment by pq** [comment]
Tracking analyzer fix in https://github.com/dart-lang/sdk/issues/52868.

**Comment by srawlins** [comment]
Fixed by https://github.com/dart-lang/sdk/commit/4c30a8990418670fab31c073394d0d19f70bda7e
